elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java

```diff
@@ -45,53 +56,195 @@
+        queryResultObj.setRecordCount(recCount);
+
+        if (resultFormatType == ResultFormatType.CSV) {
+            observableResult = convertJsonToCSV(responseBody);
+        }
+
+        if (isDownload) {
+            queryObj.setResult(queryResultObj);
+            resultStorageEngine.storeResults(queryObj, observableResult);
+            // TODO: Generate URL and set in responsebody
+            queryResultObj.setResponseBody("URL to be generated");
+        } else if (resultFormatType == ResultFormatType.CSV) {
+            queryResultObj.setResponseBody(mergeObservable(observableResult));
+        } else {
+            queryResultObj.setResponseBody(responseBody);
+        }
+
         return queryResultObj;
     }
 
+    /**
+     * This method calculates the number of records from the response.
+     * @param response is the ElideResponse
+     * @return The recordCount
+     * @throws IOException Exception thrown by JsonPath
+     */
+    protected Integer calculateRecordCount(ElideResponse response, QueryType queryType)
+            throws IOException {
+        Integer count = null;
+        if (response.getResponseCode() == 200) {
+            if (queryType.equals(QueryType.GRAPHQL_V1_0)) {
+                List<Integer> countList = JsonPath.read(response.getBody(), "$..edges.length()");
+                count = countList.size() > 0 ? countList.get(0) : 0;
+            } else if (queryType.equals(QueryType.JSONAPI_V1_0)) {
+                count = JsonPath.read(response.getBody(), "$.data.length()");
+            }
+        }
+        return count;
+    }
+
+    /**
+     * This method checks if the json string is error message.
+     * @param jsonStr is the response.getBody() we get from the response
+     * @return is the message an error message
+     */
+    protected boolean checkJsonStrErrorMessage(String jsonStr) {
+        boolean isError = false;
+        try {
+            isError = (Integer) JsonPath.read(jsonStr, "$.errors.length()") >= 1;
+        } catch (PathNotFoundException e) {
+            //ignore when not an error message
+            log.trace(e.getMessage());
+        }
+        return isError;
+    }
+
+    /**
+     * This method converts the JSON response to a CSV response type.
+     * @param jsonStr is the response.getBody() of the query
+     * @return returns an Observable of string which is in CSV format
+     * @throws IllegalStateException Exception thrown
+     */
+    protected Observable<String> convertJsonToCSV(String jsonStr) {
+        if (jsonStr == null) {
+            return null;
+        }
+        JFlat flatMe = new JFlat(jsonStr);
+        List<Object[]> json2csv;
+
+        try {
+            json2csv = flatMe.json2Sheet().headerSeparator("_").getJsonAsSheet();
+        } catch (Exception e) {
+            log.debug("Exception while converting to CSV: {}", e.getMessage());
+            throw new IllegalStateException(e);
+        }
+
+        return Observable.using(
```

Comment:
This doesn't look right to me.   We shouldn't be handling the entire response body here and convert it to CSV.  We should only convert one row at a time.